File: examples/primitives/python/src/templates/stack_pygen.py
from __future__ import annotations
from typing import *
import examples.templates.stack as pybind_stack_pygen
from bind_utils import *
from metaclass import *
from simple import *


class StackProject(metaclass=OriginalMethodsMetaclass):
    """"""
    
    @bind
    def __init__(self):
        pass

    @bind
    def __init__(self, st: Project):
        pass
    
    @bind
    def push(self, item: Project) -> None:
        """"""
        pass

    @bind
    def pop(self) -> None:
        """"""
        pass

    @bind
    def top(self) -> Project:
        """"""
        pass

    @bind
    def empty(self) -> bool:
        """"""
        pass


class StackRoot(metaclass=OriginalMethodsMetaclass):
    """"""
    
    @bind
    def __init__(self):
        pass

    @bind
    def __init__(self, st: Root):
        pass
    
    @bind
    def push(self, item: Root) -> None:
        """"""
        pass

    @bind
    def pop(self) -> None:
        """"""
        pass

    @bind
    def top(self) -> Root:
        """"""
        pass

    @bind
    def empty(self) -> bool:
        """"""
        pass

File: examples/primitives/python/wrappers/templates/stack_pygen.cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>

#include "cxx/templates/stack.hpp"
#include "cxx/simple/project.hpp"
#include "cxx/simple/root.hpp"

namespace py = pybind11;

void bindStackProject(py::module& m) {
    using namespace iegen::example;
    py::class_<iegen::example::Stack<iegen::example::Project>> stackProject(m, "StackProject");

    stackProject.def(py::init<>());

    stackProject.def(py::init<iegen::example::Project &>(), py::arg("st"));
    stackProject.def("push", &iegen::example::Stack<iegen::example::Project>::push, py::arg("item"));
    stackProject.def("pop", &iegen::example::Stack<iegen::example::Project>::pop);
    stackProject.def("top", &iegen::example::Stack<iegen::example::Project>::top);
    stackProject.def("empty", &iegen::example::Stack<iegen::example::Project>::empty);
}
void bindStackRoot(py::module& m) {
    using namespace iegen::example;
    py::class_<iegen::example::Stack<iegen::example::Root>> stackRoot(m, "StackRoot");

    stackRoot.def(py::init<>());

    stackRoot.def(py::init<iegen::example::Root &>(), py::arg("st"));
    stackRoot.def("push", &iegen::example::Stack<iegen::example::Root>::push, py::arg("item"));
    stackRoot.def("pop", &iegen::example::Stack<iegen::example::Root>::pop);
    stackRoot.def("top", &iegen::example::Stack<iegen::example::Root>::top);
    stackRoot.def("empty", &iegen::example::Stack<iegen::example::Root>::empty);
}